proposals/NNNN-extension-macros.md

```diff
@@ -0,0 +1,164 @@
+# Generalize `conformance` macros as `extension` macros
+
+* Proposal: [SE-NNNN](NNNN-extension-macros.md)
+* Authors: [Holly Borla](https://github.com/hborla)
+* Review Manager: [John McCall](https://github.com/rjmccall)
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#66967](https://github.com/apple/swift/pull/66967), [apple/swift-syntax#1859](https://github.com/apple/swift-syntax/pull/1859)
+* Review: ([pitch](https://forums.swift.org/t/pitch-generalize-conformance-macros-as-extension-macros/65653))
+
+
+## Introduction
+
+This proposal generalizes the `conformance` macro role as an `extension` macro role that can add a member list to an extension in addition to a protocol and `where` clause.
+
+## Motivation
+
+[SE-0389: Attached Macros](0389-attached-macros.md) introduced conformance macros, which expand to a conformance with a `where` clause written in an extension on the type the macro is attached to:
+
+```swift
+@attached(conformance)
+macro AddEquatable() = #externalMacro(...)
+
+@AddEquatable
+struct S {}
+
+// expands to
+extension S: Equatable {}
+```
+
+However, the `conformance` macro role is extremely limited on its own. They _only_ have the ability to return a protocol name and the syntax for a `where` clause. If the protocol conformance requires members -- as most protocol conformances do -- those must be added through a separate `member` macro role.
+
+More importantly, conformance macros are the only way for a macro to expand to an extension on the annotated type. The inability to add members in an extension of a type rather than the primary declaration is a serious limitation of the macro system, because extensions have several important semantic implications, including (but not limited to):
+
+* Protocols can only provide default implementations of requirements in extensions
+* Initializers added in an extension of a type do not suppress the compiler-synthesized initializers
+* Computed properties and methods in protocol and class extensions do not participate in dynamic dispatch
+
+Extensions also have stylistic benefits. Code inside an extension will share the generic requirements on the extension itself rather than repeating the generic requirement on every method, and implementing conformance requirements in an extension is a common practice in Swift.
+
+## Proposed solution
+
+This proposal removes the `conformance` macro role in favor or an `extension` macro role. An `extension` macro role can be used with the `@attached` macro attribute, and it can add a conformance, a where clause, and a member list in an extension on the type the macro is attached to:
```

Comment:
```suggestion
This proposal removes the `conformance` macro role in favor of an `extension` macro role. An `extension` macro role can be used with the `@attached` macro attribute, and it can add a conformance, a `where` clause, and a member list in an extension on the type the macro is attached to:
```